MSC2654: Unread counts

babolivier opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-doc/blob/babolivier/unread_counts/proposals/2654-unread-counts.md)

Synapse implementation: https://github.com/matrix-org/synapse/pull/8059
Client: https://github.com/quotient-im/libQuotient/pull/521

babolivier committed:
Add placeholder proposal

babolivier committed:
Use attributed MSC number

richvdh commented:
(to replace #2625)

babolivier committed:
Proposal

richvdh commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,90 @@
+* include a `m.relates_to` parameter in its content that includes a `rel_type`
+  parameter, which value is `m.replace`. This matches the definition of an edit
+  in [MSC1849](https://github.com/matrix-org/matrix-doc/pull/1849).
+* include a `msgtype` parameter in its content which value is `m.notice`.
```
per https://docs.google.com/document/d/1pjYrGokTQGDZQLtagZvG4dY67RsjUdFDfpOzSiWtOBg/edit?disco=AAAAJvIgsqM: probably needs to check the event type too.

richvdh commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,90 @@
+
+Homeservers currently send clients counts for missed notifications (and
+highlights) in responses to `/sync` requests, which allows them to display that
+information to clients. However, when it comes to unread messages, it doesn't
```
> it doesn't provide a count

what is "it" in this sentence?

I think it would be clearer to reverse this sentence:

> However, no count is provided for unread messages.

richvdh commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,90 @@
+messages they've missed since they last read a given room.
+
+This MSC is an alternative to
+[MSC2625](https://github.com/matrix-org/matrix-doc/pull/2625) which doesn't use
```
it kinda reads like you're saying that MSC2625 doesn't use push rules. A comma helps a bit:

```suggestion
[MSC2625](https://github.com/matrix-org/matrix-doc/pull/2625), which doesn't use
```

richvdh reviewed (dismissed):
looks good to me.

I think we also need a proposal for how to specify that certain rooms should be excluded from the badge count in `/_matrix/push/v1/notify`.

babolivier committed:
Incorporate review

babolivier commented:
> I think we also need a proposal for how to specify that certain rooms should be excluded from the badge count in /_matrix/push/v1/notify.

I assume you mean a separate MSC rather than an addition to this MSC?

richvdh commented:
> I assume you mean a separate MSC rather than an addition to this MSC?

well no, I rather think it ought to be part of this one. See also https://docs.google.com/document/d/1pjYrGokTQGDZQLtagZvG4dY67RsjUdFDfpOzSiWtOBg/edit#

turt2live commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,91 @@
+The value of this parameter is a count of every unread event (see below for a
+list of criteria an event needs to match to be marked unread) since the latest
+read receipt (i.e. the latest `m.read` read marker) for the given user in the
+given room. If no read receipt exists for this user in this room, then the value
+of the parameter is a count of every unread event since that user joined it.
```
A read receipt should always be present as we implicitly have a read receipt for the user's last sent event. We apparently haven't gotten around to specifying that though: https://github.com/matrix-org/matrix-doc/issues/569

babolivier replied:
While implementing that MSC, I realised that wasn't the case (at least Synapse doesn't store one in this case). As you said, it is currently not a specified behaviour, so I believe this mention belongs in the MSC.

turt2live commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,91 @@
+An event should be counted in the unread count only if it matches at least one
+of the following criteria:
+
+* include a non-empty `body` parameter in its content
```
do we care if it's not a string?

richvdh replied:
I vote: "Yes". if we're checking it's non-empty, we should check it's a string too.

turt2live commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,91 @@
+
+* include a non-empty `body` parameter in its content
+* be a state event which type is one of:
+    * `m.room.name`
```
why name and not avatar? why not pinned events? why these 3 specifically?

ara4n replied:
i suspect we should have `m.room.avatar` too, it was an oversight.

pinned events aren't new messages (unlike a new avatar, or a new name), so i don't think they should count.

The distinction i had in mind were things which tell the user that new first-class information has been added to the room (i.e. name, topic, avatar, new access rules, etc) as opposed to other random state.  But agreed it's quite fuzzy and subjective.  I also forget where this list came from originally, but I'm not sure I was intending it to be exhaustive.

richvdh replied:
@babolivier can we add `m.room.avatar` here?

babolivier replied:
Yes, I'm planning on updating this MSC to answer review comments later today.

Cadair replied:
Power levels feels like a weird inclusion to me. Unlike the other properties of the room (name, topic etc) it's unlikely to be new information for me to read / process most of the time. Also do PL changes cause a room to be shown with new events in most clients at the moment (using mentions only push rules etc)?

turt2live commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,91 @@
+proposal.
+
+
+## Potential issues
```
This MSC can result in a count mismatch for clients: notifications could say 12 notified messages but the server says 3 unread messages. Which one is the client to prefer/believe?

richvdh replied:
if it has access to the full /sync data, it should probably use that. the `notifications` update is to keep the badge count up-to-date on clients that aren't syncing.

turt2live commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,91 @@
+    * `m.room.name`
+    * `m.room.topic`
+    * `m.room.power_levels`
+* be an encrypted message (i.e. have the `m.room.encrypted` type)
```
Should the client re-execute these rules itself like they do for push rules once the event is decrypted?

cloudrac3r replied:
Did you find out the answer?

turt2live reviewed (dismissed):
Overall this feels fine, though some rationale for the decisions would be appreciated.

richvdh commented:
thanks for the detailed review and helping to spot the edge-cases we forgot, @turt2live !

babolivier committed:
Incorporate review

babolivier committed:
Mention that users shouldn't get their unread count up'd by their own messages

tulir commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,106 @@
+of the following criteria:
+
+* include a non-empty `body` parameter, which value is a string, in its content
+* be a state event which type is one of:
+    * `m.room.name`
+    * `m.room.topic`
+    * `m.room.power_levels`
+    * `m.room.avatar`
```
Why `m.room.power_levels` and why not `m.room.pinned_events` and `m.room.tombstone`?

richvdh replied:
(and removing `m.room.power_levels`)

babolivier replied:
ftr pinned events have already been discussed here: https://github.com/matrix-org/matrix-doc/pull/2654#discussion_r448270153

tulir commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,106 @@
+
+* include a `m.relates_to` parameter in its content that includes a `rel_type`
+  parameter, which value is `m.replace`. This matches the definition of an edit
+  in [MSC1849](https://github.com/matrix-org/matrix-doc/pull/1849).
```
```suggestion
  in [MSC2676](https://github.com/matrix-org/matrix-doc/pull/2676).
```

tulir commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,106 @@
+## Potential issues
+
+This MSC mentions edits, which are specified in
+[MSC1849](https://github.com/matrix-org/matrix-doc/pull/1849). Therefore, it is
+unclear whether this MSC can be merged before MSC1849.
```
```suggestion
[MSC2676](https://github.com/matrix-org/matrix-doc/pull/2676). Therefore, it is
unclear whether this MSC can be merged before MSC2676.
```

babolivier committed:
s/MSC1849/MSC2676/

Co-authored-by: Tulir Asokan <tulir@maunium.net>

babolivier committed:
Update proposals/2654-unread-counts.md

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

babolivier committed:
Update the list of state events to count unread

Added tombstone, removed power levels

babolivier committed:
Merge branch 'babolivier/unread_counts' of github.com:matrix-org/matrix-doc into babolivier/unread_counts

babolivier committed:
Remove implementation notes on /notify

This actually leads to a pretty confusing UX. We might revert that in the future, but probably not as part of this MSC.

KitsuneRal commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,96 @@
+* include a `m.relates_to` parameter in its content that includes a `rel_type`
+  parameter, which value is `m.replace`. This matches the definition of an edit
+  in [MSC2676](https://github.com/matrix-org/matrix-doc/pull/2676).
+* have the type `m.room.message`, no `state_key`, and include a `msgtype`
+  parameter in its content which value is `m.notice`.
```
What if an `m.notice` event _has_ `state_key`? Officially that would make it a (spec-incompliant) state event; and since it's not matching any type from the list above, it's not icluded in the unread count anyway. Seems like "no `state_key`" condition is superfluous here.

KitsuneRal replied:
In addition to `m.notice`, it's worth excluding `m.key.verification.*` messages (this can only be done on the client side, of course).

KitsuneRal commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,96 @@
+list of criteria an event needs to match to be marked unread) since the latest
+read receipt (i.e. the latest `m.read` read marker) for the given user in the
+given room. If no read receipt exists for this user in this room, then the value
+of the parameter is a count of every unread event since that user joined it.
```
Considering the case when the user joined, left, and joined again, all without ever reading the timeline. Should the number of unread messages be counted from the first or from the latest join?

zecakeh replied:
Since there are no implicit read receipts, shouldn't that be the number of unread messages since the last event sent by the user, or something like that?

KitsuneRal reviewed (approved):
This looks quite useful to me: specifying the way events are counted would help clients and homeservers to do it in a uniform way (and actually, Quotient devised event counting logic that served exactly this purpose but used an inferior algorithm). I have a couple of nitpicks below; but I think this is generally good to go.

KitsuneRal commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,96 @@
+    * `m.room.name`
+    * `m.room.topic`
+    * `m.room.avatar`
+    * `m.room.tombstone`
```
https://github.com/matrix-org/matrix-react-sdk/blob/84af29a7f1a5e4d4a8f3408f0f741293f10be010/src/components/views/rooms/EventTile.tsx#L1186 (as called from https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/Unread.ts#L32) has another interesting addition - it counts an event as unread if it's `m.room.create` with non-empty `predecessor`. To be honest, I don't quite understand the rationale behind that and am just documenting the case.

KitsuneRal reviewed (changes requested):
After studying further the current logic in matrix-react-sdk, there might be a couple more things worth adding to the criteria.

babolivier committed:
Add mention about highlight related limitation

erikjohnston commented:
What's the status of this? There is currently code in Synapse to make this work that is on the hot path, and we should rip it out if we're not going to finish this any time soon.

babolivier commented:
> What's the status of this? There is currently code in Synapse to make this work that is on the hot path, and we should rip it out if we're not going to finish this any time soon.

I think the next steps are addressing the concerns in threads and asking for FCP to be proposed. I'm unlikely to get bandwidth to do that anytime soon though.

erikjohnston commented:
Is this actually being driven by something? If it's not a feature that people are asking for then I'm not sure there's much point finishing it?

babolivier commented:
> Is this actually being driven by something? If it's not a feature that people are asking for then I'm not sure there's much point finishing it?

It was driven by the notification rework that was underway at the time, but isn't anymore. I have occasionally seen people use this feature, but rarely and I believe this is a pretty niche feature, so I don't mind if we decide to abandon it.

KitsuneRal commented:
Upcoming libQuotient 0.7 also has it implemented (driving my feedback here).

richvdh commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,103 @@
+
+## Potential issues
+
+This MSC mentions edits, which are specified in
+[MSC2676](https://github.com/matrix-org/matrix-doc/pull/2676). Therefore, it is
+unclear whether this MSC can be merged before MSC2676.
```
MSC2676 is now merged.

richvdh commented:
I don't think it's reasonable to keep this "experimental" code in synapse's mainline indefinitely; we either need to commit to speccing it properly, or to removing it. 

It's worth noting that enabling the option in Synapse has a substantial performance impact, so making it part of the spec would require us to find a solution to that.

Given that the author of this MSC has said that he is no longer interested in pursuing it, I think that it will need a volunteer to take it on; otherwise it is likely to be closed as "abandoned".

So, any volunteers to address the comments and think about how to solve the performance problems?

KitsuneRal commented:
I'm willing to take this over - is there a write-up of the performance issues somewhere?

clokep commented:
> I'm willing to take this over - is there a write-up of the performance issues somewhere?

Not really, no. matrix-org/synapse#13694 is the PR of interest. Happy to chat over there (or we can file a separate issue on Synapse referencing that if you'd like).

tezlm commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,103 @@
+
+This MSC proposes the addition of an `unread_count` parameter in the `Joined
+Room` section of
+[`/sync`](https://matrix.org/docs/spec/client_server/r0.6.1#get-matrix-client-r0-sync)
```
Did you mean `unread_notifications.unread_count`?

bretthowell714-source commented on proposals/2654-unread-counts.md:
```diff
@@ -0,0 +1,103 @@
+since they last read a given room.
+
+This MSC is an alternative to
+[MSC2625](https://github.com/matrix-org/matrix-doc/pull/2625), which doesn't use
```
Thanks
